### PR TITLE
Issue 8185: Allow square brackets in spec files

### PR DESCRIPTION
**Pull request by perillai**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)
This is to fix [Issue 8185](https://github.com/webdriverio/webdriverio/issues/8185). Used a suggestion provided in the GitHub repo of Glob. Works for all directories below the 'test' folder.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

<a href="https://gitpod.io/#https://github.com/webdriverio/webdriverio/pull/8655"></a>

### CODE REVIEW

**Comment by perillai** [comment]
Getting 'is-stream' issue while running unit tests. Currently, is-stream v3 wants import to be used instead of require. But, all across our project, I am able to see v2, but the npm 'tempy' has 'is-stream' 3 as dependency. Because of this, wherever we require tempy, it creates "SyntaxError: Named export 'isStream' not found. The requested module 'is-stream' is a CommonJS module, which may not support all module.exports as named exports." Currently, two tests under wdio-allure-reporter is failing because of this. Is this an issue faced by me alone? or do others face this? If so, can we rollback the tempy version to 1?

**Comment by christian-bromann** [comment]
Can we add a test for this?

> Getting 'is-stream' issue while running unit tests.

Let's log this as a different issue.

**Comment by erwinheitzman** [comment]
Why do we need a double escape in the string? (I mean the `\\[` and `\\]`)

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 1]
```diff
@@ -348,6 +348,7 @@ export default class ConfigParser {
                 log.error('Unexpected depth of hierarchical arrays')
             } else {
                 pattern = pattern.toString().replace(/\\/g, '/')
+                pattern = pattern.toString().replace(/[[\]]/g, '\\$&')
```
Comment (erwinheitzman) [inline packages/wdio-config/src/lib/ConfigParser.ts:351]: I think you missed the escape for the opening bracket
```suggestion
                pattern = pattern.toString().replace(/[\[\]]/g, '\\$&')
```

**Comment by erwinheitzman** [comment]
> Why do we need a double escape in the string? (I mean the `\\[` and `\\]`)

Apparently this is needed, I just checked as I found it odd. It's been a while since I worked with backslashes in strings, you noticed? haha

Thanks for the PR!

File: packages/wdio-config/src/lib/ConfigParser.ts (hunk 1 above)
Comment (perillai) [reply packages/wdio-config/src/lib/ConfigParser.ts:351]: > Can we add a test for this?
> 
> > Getting 'is-stream' issue while running unit tests.
> 
> Let's log this as a different issue.

What do you think can be the fix? Downgrading tempy?

Comment (christian-bromann) [reply packages/wdio-config/src/lib/ConfigParser.ts:351]: let's discuss this in #8671

**Comment by christian-bromann** [comment]
@perillai please revert the latest commits, they are irrelevant to the issue your are solving originally. Furthermore the current `main` branch supports ESM and should have no issue with `tempy`.

**Comment by christian-bromann** [comment]
The `v7` branch which is what currently is released uses v1 of tempy: https://github.com/webdriverio/webdriverio/blob/v7/package.json#L103

**Comment by perillai** [comment]
> @perillai please revert the latest commits, they are irrelevant to the issue your are solving originally. Furthermore the current `main` branch supports ESM and should have no issue with `tempy`.

I have reverted those commits.

**Review by christian-bromann — COMMENTED** [review]
Mind reverting changes in `package-lock.json` as well, addressing @erwinheitzman comment and add a unit test?

**Comment by perillai** [comment]
> Mind reverting changes in `package-lock.json` as well, addressing @erwinheitzman comment and add a unit test?

I have reverted changes in package-lock. And @erwinheitzman has accepted that the committed code looks fine. 
Sorry to say this, unit test is totally new to me, as I am only used to doing automated e2e tests. I was trying to add unit test for this and still unable to understand how to add one. Can I add unit test for this later?

**Comment by christian-bromann** [comment]
> Can I add unit test for this later?

Unfortunately not. Unit tests need to be added [here](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-config/tests/configparser.test.ts). You can run them via:

```sh
npx vitest packages/wdio-config/tests/configparser.test.ts
```

You can find further docs on writing unit tests here: https://vitest.dev/

**Comment by christian-bromann** [comment]
@perillai any updates on this?

**Comment by perillai** [comment]
> @perillai any updates on this?

Sorry for the delay. Figured out how to add unit test using vitest. Hope it looks fine!

**Comment by christian-bromann** [comment]
Closing in favor of #9004

I don't think this solution does work as it only removes the brackets. @vmwhereqa has worked on a different solution while this PR became stale. Unfortunately we didn't communicated this clearly so I was a but surprised that you picked it up again. Happy to re-assign a different issue to you.

Anyway, thanks for the work on this it helped @vmwhereqa to understand the problem and come up with a solution as well.

**Pull request closed without merging**